<!DOCTYPE html><html><div class="item-title">
        Item 195
      </div> <div class="item-details"><div><b>git_comments:</b> <ol><li><div>
                * Shutdown hook thread to ensure deletion of the storage directory. 
              </div></li><li><div><div><b>comment:</b>  Remove shutdown hook to prevent resource leaks
                </div><div><b>label:</b> code-design
                </div></div></li><li><div>
                * Shutdown hook thread to ensure deletion of the storage directory. 
              </div></li><li><div>
                 Add JVM shutdown hook to call shutdown of service
              </div></li></ol></div><div><b>git_commits:</b> <ol><li><div><div><b>summary:</b> [FLINK-1492] Fix exceptions on blob store shutdown
                </div><div><b>message:</b> [FLINK-1492] Fix exceptions on blob store shutdown

This closes #376

                </div></div></li></ol></div><div><b>github_issues:</b> <ol></ol></div><div><b>github_issues_comments:</b> <ol></ol></div><div><b>github_pulls:</b> <ol><li><div><div><b>title:</b> [FLINK-1492] Fix exceptions on blob store shutdown
                </div><div><b>body:</b> 
                </div></div></li><li><div><div><b>title:</b> [FLINK-1492] Fix exceptions on blob store shutdown
                </div><div><b>body:</b> 
                </div></div></li><li><div><div><b>title:</b> [FLINK-1492] Fix exceptions on blob store shutdown
                </div><div><b>body:</b> 
                </div></div></li><li><div><div><b>title:</b> [FLINK-1492] Fix exceptions on blob store shutdown
                </div><div><b>body:</b> 
                </div></div></li><li><div><div><b>title:</b> [FLINK-1492] Fix exceptions on blob store shutdown
                </div><div><b>body:</b> 
                </div></div></li><li><div><div><b>title:</b> [FLINK-1492] Fix exceptions on blob store shutdown
                </div><div><b>body:</b> 
                </div></div></li><li><div><div><b>title:</b> [FLINK-1492] Fix exceptions on blob store shutdown
                </div><div><b>body:</b> 
                </div></div></li><li><div><div><b>title:</b> [FLINK-1492] Fix exceptions on blob store shutdown
                </div><div><b>body:</b> 
                </div><div><b>label:</b> code-design
                </div></div></li><li><div><div><b>title:</b> [FLINK-1492] Fix exceptions on blob store shutdown
                </div><div><b>body:</b> 
                </div></div></li><li><div><div><b>title:</b> [FLINK-1492] Fix exceptions on blob store shutdown
                </div><div><b>body:</b> 
                </div></div></li><li><div><div><b>title:</b> [FLINK-1492] Fix exceptions on blob store shutdown
                </div><div><b>body:</b> 
                </div></div></li><li><div><div><b>title:</b> [FLINK-1492] Fix exceptions on blob store shutdown
                </div><div><b>body:</b> 
                </div></div></li><li><div><div><b>title:</b> [FLINK-1492] Fix exceptions on blob store shutdown
                </div><div><b>body:</b> 
                </div></div></li><li><div><div><b>title:</b> [FLINK-1492] Fix exceptions on blob store shutdown
                </div><div><b>body:</b> 
                </div></div></li></ol></div><div><b>github_pulls_comments:</b> <ol><li><div>
                This change is also in 0.8 so do we need to apply the fix there as well for the upcoming 0.8.1 release?

              </div></li><li><div>
                Yes, if it is finally OK.

              </div></li><li><div>
                Looks good.

              </div></li><li><div>
                LGTM except for my single question.

              </div></li><li><div>
                I'll merge the change.

              </div></li><li><div>
                OK, thanks. Just a reminder: we need to include this in 0.8.1 as well.

              </div></li><li><div>
                I still get 

```
05:26:08,841 ERROR org.apache.flink.runtime.blob.BlobServer  - Error during shutdown of blob service via JVM shutdown hook: Shutdown in progress
java.lang.IllegalStateException: Shutdown in progress
at java.lang.ApplicationShutdownHooks.remove(ApplicationShutdownHooks.java:82)
at java.lang.Runtime.removeShutdownHook(Runtime.java:239)
at org.apache.flink.runtime.blob.BlobServer.shutdown(BlobServer.java:220)
at org.apache.flink.runtime.blob.BlobUtils$1.run(BlobUtils.java:210)
at java.lang.Thread.run(Thread.java:745)
```

On the last night's master.

              </div></li><li><div><div><b>body:</b> OK, thanks for reporting this. @StephanEwen has some pending changes to the blob manager and he will look into it as well.

It is not allowed to remove the shutdown hook when a shutdown is already in progress.

                </div><div><b>label:</b> code-design
                </div></div></li><li><div>
                The problem is totally deterministic when shutting down the job/task managers.

I've fixed it here: https://github.com/uce/incubator-flink/tree/flink-1492-fix_exceptions_really

@StephanEwen, if this is blocking the release vote, we should just merge my change and not wait for your changes in master, which touch the blob store.

              </div></li></ol></div><div><b>github_pulls_reviews:</b> <ol><li><div>
                What does this join do? I'm at a loss here.

              </div></li><li><div>
                It's from the old code and I am not sure if it really needs to stay, but it ensures that the BlobServer thread really finishes when calling the shutdown method (BlobServer is a Thread and because the join is called from outside of the run method it waits for the BlobServer thread to finish).

              </div></li><li><div>
                Ah of course. Thanks for the clarification.

              </div></li></ol></div><div><b>jira_issues:</b> <ol><li><div><div><b>summary:</b> Exceptions on shutdown concerning BLOB store cleanup
                </div><div><b>description:</b> The following stack traces occur not every time, but frequently.

{code}
java.lang.IllegalArgumentException: /tmp/blobStore-7a89856a-47f9-45d6-b88b-981a3eff1982 does not exist
	at org.apache.commons.io.FileUtils.cleanDirectory(FileUtils.java:1637)
	at org.apache.commons.io.FileUtils.deleteDirectory(FileUtils.java:1535)
	at org.apache.flink.runtime.blob.BlobServer.shutdown(BlobServer.java:213)
	at org.apache.flink.runtime.execution.librarycache.BlobLibraryCacheManager.shutdown(BlobLibraryCacheManager.java:171)
	at org.apache.flink.runtime.jobmanager.JobManager.postStop(JobManager.scala:136)
	at akka.actor.Actor$class.aroundPostStop(Actor.scala:475)
	at org.apache.flink.runtime.jobmanager.JobManager.aroundPostStop(JobManager.scala:80)
	at akka.actor.dungeon.FaultHandling$class.akka$actor$dungeon$FaultHandling$$finishTerminate(FaultHandling.scala:210)
	at akka.actor.dungeon.FaultHandling$class.handleChildTerminated(FaultHandling.scala:292)
	at akka.actor.ActorCell.handleChildTerminated(ActorCell.scala:369)
	at akka.actor.dungeon.DeathWatch$class.watchedActorTerminated(DeathWatch.scala:63)
	at akka.actor.ActorCell.watchedActorTerminated(ActorCell.scala:369)
	at akka.actor.ActorCell.invokeAll$1(ActorCell.scala:455)
	at akka.actor.ActorCell.systemInvoke(ActorCell.scala:478)
	at akka.dispatch.Mailbox.processAllSystemMessages(Mailbox.scala:279)
	at akka.dispatch.Mailbox.run(Mailbox.scala:220)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:231)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.pollAndExecAll(ForkJoinPool.java:1253)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1346)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)

15:16:15,350 ERROR org.apache.flink.test.util.ForkableFlinkMiniCluster$$anonfun$startTaskManager$1$$anon$1  - LibraryCacheManager did not shutdown properly.
java.io.IOException: Unable to delete file: /tmp/blobStore-e2619536-fb7c-452a-8639-487a074d1582/cache/blob_ff74895f7bdeeaa3bd70b6932beed143048bb4c7
	at org.apache.commons.io.FileUtils.forceDelete(FileUtils.java:2279)
	at org.apache.commons.io.FileUtils.cleanDirectory(FileUtils.java:1653)
	at org.apache.commons.io.FileUtils.deleteDirectory(FileUtils.java:1535)
	at org.apache.commons.io.FileUtils.forceDelete(FileUtils.java:2270)
	at org.apache.commons.io.FileUtils.cleanDirectory(FileUtils.java:1653)
	at org.apache.commons.io.FileUtils.deleteDirectory(FileUtils.java:1535)
	at org.apache.flink.runtime.blob.BlobCache.shutdown(BlobCache.java:159)
	at org.apache.flink.runtime.execution.librarycache.BlobLibraryCacheManager.shutdown(BlobLibraryCacheManager.java:171)
	at org.apache.flink.runtime.taskmanager.TaskManager.postStop(TaskManager.scala:173)
	at akka.actor.Actor$class.aroundPostStop(Actor.scala:475)
	at org.apache.flink.runtime.taskmanager.TaskManager.aroundPostStop(TaskManager.scala:86)
	at akka.actor.dungeon.FaultHandling$class.akka$actor$dungeon$FaultHandling$$finishTerminate(FaultHandling.scala:210)
	at akka.actor.dungeon.FaultHandling$class.terminate(FaultHandling.scala:172)
	at akka.actor.ActorCell.terminate(ActorCell.scala:369)
	at akka.actor.ActorCell.invokeAll$1(ActorCell.scala:462)
	at akka.actor.ActorCell.systemInvoke(ActorCell.scala:478)
	at akka.dispatch.Mailbox.processAllSystemMessages(Mailbox.scala:279)
	at akka.dispatch.Mailbox.run(Mailbox.scala:220)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:231)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.pollAndExecAll(ForkJoinPool.java:1253)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1346)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)

15:16:15,345 ERROR org.apache.flink.runtime.blob.BlobCache                       - Error deleting directory /tmp/blobStore-4313349e-8a58-4683-9fd0-3d2c52be1864 during JVM shutdown: /tmp/blobStore-4313349e-8a58-4683-9fd0-3d2c52be1864 does not exist
java.lang.IllegalArgumentException: /tmp/blobStore-4313349e-8a58-4683-9fd0-3d2c52be1864 does not exist
	at org.apache.commons.io.FileUtils.cleanDirectory(FileUtils.java:1637)
	at org.apache.commons.io.FileUtils.deleteDirectory(FileUtils.java:1535)
	at org.apache.flink.runtime.blob.BlobUtils$1.run(BlobUtils.java:210)
	at java.lang.Thread.run(Thread.java:745)

{code}
                </div></div></li></ol></div><div><b>jira_issues_comments:</b> <ol><li><div><div><b>body:</b> The current solution is a bit hacky. Right now, we see multiple cleanups I think one is on graceful shutdown of the task manager (through observation of job manager death) and then through the shutdown hook.

I think the right solution is to not simply let the shutdown hook delete the directory, but to have the shutdown hook trigger call a "shutdown" on the Blob manager.
The shutdown should also make sure it occurs only once, so it does not happen through both the task manager shutdown, and the shutdown hook.

It is also good practice that the blob manager should remove the shutdown hook once shutdown is called, to prevent resource leaks.

                </div><div><b>label:</b> code-design
                </div></div></li><li><div>
                GitHub user uce opened a pull request:

    https://github.com/apache/flink/pull/376

    [FLINK-1492] Fix exceptions on blob store shutdown

    

You can merge this pull request into a Git repository by running:

    $ git pull https://github.com/uce/incubator-flink flink-1492-proper_shutdown_hook

Alternatively you can review and apply these changes as the patch at:

    https://github.com/apache/flink/pull/376.patch

To close this pull request, make a commit to your master/trunk branch
with (at least) the following in the commit message:

    This closes #376
    
----
commit f29132ca9fb549d47a29322695f467996ed727a4
Author: Ufuk Celebi &lt;uce@apache.org&gt;
Date:   2015-02-09T10:44:47Z

    [FLINK-1492] Fix exceptions on blob store shutdown

----

              </div></li><li><div>
                Github user rmetzger commented on the pull request:

    https://github.com/apache/flink/pull/376#issuecomment-73500870
  
    This change is also in 0.8 so do we need to apply the fix there as well for the upcoming 0.8.1 release?

              </div></li><li><div>
                Github user uce commented on the pull request:

    https://github.com/apache/flink/pull/376#issuecomment-73500986
  
    Yes, if it is finally OK.

              </div></li><li><div>
                Github user rmetzger commented on the pull request:

    https://github.com/apache/flink/pull/376#issuecomment-73667317
  
    Looks good.

              </div></li><li><div>
                Github user tillrohrmann commented on a diff in the pull request:

    https://github.com/apache/flink/pull/376#discussion_r24402830
  
    --- Diff: flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobServer.java ---
    @@ -196,23 +199,28 @@ public void run() {
     	 */
     	@Override
     	public void shutdown() throws IOException {
    -
    -		this.shutdownRequested = true;
    -		try {
    -			this.serverSocket.close();
    -		} catch (IOException ioe) {
    +		if (shutdownRequested.compareAndSet(false, true)) {
    +			try {
    +				this.serverSocket.close();
    +			}
    +			catch (IOException ioe) {
     				LOG.debug("Error while closing the server socket.", ioe);
    -		}
    -		try {
    -			join();
    -		} catch (InterruptedException ie) {
    -			LOG.debug("Error while waiting for this thread to die.", ie);
    -		}
    +			}
    +			try {
    +				join();
    --- End diff --
    
    What does this join do? I'm at a loss here.

              </div></li><li><div>
                Github user tillrohrmann commented on the pull request:

    https://github.com/apache/flink/pull/376#issuecomment-73678581
  
    LGTM except for my single question.

              </div></li><li><div>
                Github user uce commented on a diff in the pull request:

    https://github.com/apache/flink/pull/376#discussion_r24403795
  
    --- Diff: flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobServer.java ---
    @@ -196,23 +199,28 @@ public void run() {
     	 */
     	@Override
     	public void shutdown() throws IOException {
    -
    -		this.shutdownRequested = true;
    -		try {
    -			this.serverSocket.close();
    -		} catch (IOException ioe) {
    +		if (shutdownRequested.compareAndSet(false, true)) {
    +			try {
    +				this.serverSocket.close();
    +			}
    +			catch (IOException ioe) {
     				LOG.debug("Error while closing the server socket.", ioe);
    -		}
    -		try {
    -			join();
    -		} catch (InterruptedException ie) {
    -			LOG.debug("Error while waiting for this thread to die.", ie);
    -		}
    +			}
    +			try {
    +				join();
    --- End diff --
    
    It's from the old code and I am not sure if it really needs to stay, but it ensures that the BlobServer thread really finishes when calling the shutdown method (BlobServer is a Thread and because the join is called from outside of the run method it waits for the BlobServer thread to finish).

              </div></li><li><div>
                Github user tillrohrmann commented on a diff in the pull request:

    https://github.com/apache/flink/pull/376#discussion_r24407836
  
    --- Diff: flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobServer.java ---
    @@ -196,23 +199,28 @@ public void run() {
     	 */
     	@Override
     	public void shutdown() throws IOException {
    -
    -		this.shutdownRequested = true;
    -		try {
    -			this.serverSocket.close();
    -		} catch (IOException ioe) {
    +		if (shutdownRequested.compareAndSet(false, true)) {
    +			try {
    +				this.serverSocket.close();
    +			}
    +			catch (IOException ioe) {
     				LOG.debug("Error while closing the server socket.", ioe);
    -		}
    -		try {
    -			join();
    -		} catch (InterruptedException ie) {
    -			LOG.debug("Error while waiting for this thread to die.", ie);
    -		}
    +			}
    +			try {
    +				join();
    --- End diff --
    
    Ah of course. Thanks for the clarification.

              </div></li><li><div>
                Github user rmetzger commented on the pull request:

    https://github.com/apache/flink/pull/376#issuecomment-73699086
  
    I'll merge the change.

              </div></li><li><div>
                Github user uce commented on the pull request:

    https://github.com/apache/flink/pull/376#issuecomment-73699625
  
    OK, thanks. Just a reminder: we need to include this in 0.8.1 as well.

              </div></li><li><div>
                Github user asfgit closed the pull request at:

    https://github.com/apache/flink/pull/376

              </div></li><li><div>
                Resolved for master in http://git-wip-us.apache.org/repos/asf/flink/commit/b88f909c
              </div></li><li><div>
                Resolved for 0.8.1 in https://git1-us-west.apache.org/repos/asf?p=flink.git;a=commit;h=5b420d84
              </div></li><li><div>
                Github user mxm commented on the pull request:

    https://github.com/apache/flink/pull/376#issuecomment-73855452
  
    I still get 
    
        05:26:08,841 ERROR org.apache.flink.runtime.blob.BlobServer  - Error during shutdown of blob service via JVM shutdown hook: Shutdown in progress
        java.lang.IllegalStateException: Shutdown in progress
    	at java.lang.ApplicationShutdownHooks.remove(ApplicationShutdownHooks.java:82)
    	at java.lang.Runtime.removeShutdownHook(Runtime.java:239)
    	at org.apache.flink.runtime.blob.BlobServer.shutdown(BlobServer.java:220)
    	at org.apache.flink.runtime.blob.BlobUtils$1.run(BlobUtils.java:210)
    	at java.lang.Thread.run(Thread.java:745)
    
    On the last night's master.

              </div></li><li><div>
                Github user uce commented on the pull request:

    https://github.com/apache/flink/pull/376#issuecomment-73856074
  
    OK, thanks for reporting this. @StephanEwen has some pending changes to the blob manager and he will look into it as well.
    
    It is not allowed to remove the shutdown hook when a shutdown is already in progress.

              </div></li><li><div>
                Github user uce commented on the pull request:

    https://github.com/apache/flink/pull/376#issuecomment-73859776
  
    The problem is totally deterministic when shutting down the job/task managers.
    
    I've fixed it here: https://github.com/uce/incubator-flink/tree/flink-1492-fix_exceptions_really
    
    @StephanEwen, if this is blocking the release vote, we should just merge my change and not wait for your changes in master, which touch the blob store.

              </div></li></ol></div></div></html>